### PR TITLE
Move results attributes of helix_analysis to Results

### LINKED ISSUE #3261: make AnalysisBase-derived classes use .results
## Is your feature request related to a problem? ##

With PR #3233 merged, we established a new API for AnalysisBase-derived analysis classes: results should be stored in the `.results` attribute, which is a dict that makes keys available as attributes. Any classes that have not been using results should start using it _in a backwards compatible manner_.

## Describe the solution you'd like ##

Add `.results` to all Analysis-classes in such a way that access to the old documented attributes still works for the full 2.x life cycle.

### Classes
The following classes should be upgraded (roughly in order of importance) --- raise individual issues and add them here; add more classes to be upgraded as necessary
- [x] BAT (get this done in 2.0.0 since it was added then = no deprecation needed) #3272
- [x] helix_analysis (as above needs, to get done before 2.0.0 gets released)  #3268
- [x] rms.RMSD PR #3274
- [x] rms.RMSF PR #3274
- [x] align.AlignTraj and align.AverageStructure #3278 
- [x] hydrogenbonds.HydrogenBondAnalysis #3271
- [x] contacts.ContactAnalysis PR #3264
- [x] density.DensityAnalysis PR #3263
- [x] pca.PCA #3275
- [x] dihedrals.Ramachandran PR #3266
- [x] dihedrals.Dihedrals PR #3266
- [x] dihedrals.Janin PR #3266
- [x] msd.EinsteinMSD PR #3265
- [x] rdf.InterRDF #3276
- [x] rdf.InterRDF_s #3276
- [x] hydrogenbonds.WaterBridgeAnalysis #3270
- [x] hole2.HoleAnalysis #3269
- [x] DiffusionMap.DistanceMatrix #3290 
- [x] polymer.PersistenceLength #3291
- [ ] ...

### Approach
Generally follow the approach taken for DensityAnalysis in PR #3263

#### Code
Assume that the old code in a class `ExampleAnalysis(AnalysisBase)` stored a result in attribute `foo`. 

* change to using `.results`:
  * You do _not_ need to set up `self.results = Results()` because `AnalysisBase` does it.
  * You must change any attribute assignments to be in `self.results` instead of `self`.  For example, `self.foo = data` becomes `self.results.foo = data`. 
  * Note that [`Results`](https://docs.mdanalysis.org/2.0.0-dev0/documentation_pages/analysis/base.html#MDAnalysis.analysis.base.Results) is really a dict where keys can be accessed as attributes. This can be useful when key access is more convenient than attribute access. 
  * Results may be nested if necessary.
* Required changes to keep access to `foo` backward compatible + deprecation warning:
```python
import warnings


class ExampleAnalysis(AnalysisBase):
    ...
    @property
    def foo(self):
        wmsg = ("The `foo` attribute was deprecated in MDAnalysis 2.0.0 "
                "and will be removed in MDAnalysis 3.0.0. Please use "
                "`results.foo` instead")
        warnings.warn(wmsg, DeprecationWarning)
        return self.results.foo
   ```

#### Docs
Change all examples in the docs to use the new `.results`. Pay particular attention to any ``.. attribute::` section and examples. Add a deprecation note in docs directly to the attribute:
```reST

.. attribute:: foo

   The results of the calculation are stored here.

   .. deprecated:: 2.0.0 
       This attribute will be removed in 3.0.0. Use :attr:`results.foo` instead.
```
Add docs for the results attribute with a versionchanged
```reST
.. attribute:: results.foo

   The results of the calculation are stored here.

   .. versionadded:: 2.0.0 
```

Add a version changed at the bottom of the class docs
```reST
.. versionchanged:: 2.0.0
       Results are now stored in a :class:`MDAnalysis.analysis.base.Results` instance.
```

If necessary, raise [issues for the User Guide](https://github.com/MDAnalysis/UserGuide/issues) if you already know that it needs updating, too.

#### CHANGELOG
```
Changes
  * `ExampleAnalysis` now uses the `results.foo` attribute for storing
    data. The `ExampleAnalysis.foo` attribute is now deprecated
    (Issue #xxxx)

Deprecations
  * The `foo` attribute of `analysis.example.ExampleAnalysis.foo` is now
    deprecated in favour of `results.foo`. It will be removed in 3.0.0
    (Issue #xxxx)
```

### Do nothing
Do nothing right now and update as we go along. The downside is that we then miss the opportunity to educate users right from 2.0.0 onwards to use the new API.

### Other ways to still make the attributes available

For results that can be stored by reference, it will be sufficient to just assign to `self.results.NAME` in addition to `self.NAME`:

```python
class ExampleAnalysis(base.AnalysisBase):
   ...

   def _conclude(self):
         # old
         # self.data = np.array(self._data)
         # new (remove self.data in 3.0.0)
         self.results.data = self.data = np.array(self._data)
```
This simple approach has the disadvantage that we eventually have to manually add deprecation warnings that the old attributes will be removed. 

If necessary, one could use properties and also add the deprecation warning:

```python
from MDAnalysis.lib.util import deprecate


class ExampleAnalysis(AnalysisBase):

   def _conclude(self):
        self.results.data = ...

   @property
   @deprecate(old_name="ExampleAnalysis.data", new_name="ExampleAnalysis.results.data",
            release="2.0.0", remove="3.0.0")
   def data(self):
        return self.results.data
   
   # for full compatibility there should be a setter but I am not sure if really needed
   @data.setter
   def data(self, value):
         self.results.data = value
```

## Additional context ##
* https://github.com/MDAnalysis/mdanalysis/issues/3115#issuecomment-832582333
* https://github.com/MDAnalysis/mdanalysis/issues/3115#issuecomment-832753553

### LINKED ISSUE #3267: Incorrect versionadded for helix_analysis?
The helix_analysis docstring says that it was added in version 1.0.0

https://github.com/MDAnalysis/mdanalysis/blob/2c714fccd95d8b05f807e38175557e398199b1d6/package/MDAnalysis/analysis/helix_analysis.py#L32

However, it's not (unless I'm missing something) in 1.1.1 (or at least current `master`).

### LINKED ISSUE #3268: update helix_analysis to use .results
Follow the new AnalysisBase API for the new helix_analysis and store results in `.results` instead of class attributes.

This should be done for 2.0.0 #3254 so that we don't have to start this new analysis module with deprecations in place.

See omnibus issue #3261 for background and details. See PR #3263 as an example (DensityAnalysis) but note that _no deprecations are necessary_ — just change where the results are stored and update the docs.

- [ ] use attributes in`self.results` instead of `self`
- [ ] update docs
- [ ] add an entry in CHANGELOG under Changes (for completeness)

### PR DESCRIPTION
**Pull request by lilyminium**
Fixes #3268
Fixes #3267

Changes made in this Pull Request:
 - changed versionchanged to the right number
 - Moved attributes to `results`

I did *not* add a deprecation warning because this analysis is new in 2.0.0.

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

### CODE REVIEW

File: package/MDAnalysis/analysis/helix_analysis.py [hunk 1]
```diff
@@ -418,49 +420,49 @@ def _prepare(self):
         for key, dims in self.attr_shapes.items():
             empty = [self._zeros_per_frame(
                 dims, n_positions=n) for n in n_res]
-            setattr(self, key, empty)
+            setattr(self.results, key, empty)
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/helix_analysis.py:423]: fyi: Could now be written as
```suggestion
            self.results[key] = empty
```

File: package/MDAnalysis/analysis/helix_analysis.py [hunk 2]
```diff
@@ -418,49 +420,49 @@ def _prepare(self):
...
+        self.results.global_axis = [self._zeros_per_frame((3,)) for n in n_res]
+        self.results.all_bends = [self._zeros_per_frame((n-3, n-3)) for n in n_res]
 
     def _single_frame(self):
         _f = self._frame_index
         for i, ag in enumerate(self.atomgroups):
             results = helix_analysis(ag.positions, ref_axis=self.ref_axis)
             for key, value in results.items():
-                attr = getattr(self, key)
+                attr = getattr(self.results, key)
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/helix_analysis.py:433]: fyi
```suggestion
                attr = self.results[key]
```

File: package/MDAnalysis/analysis/helix_analysis.py [hunk 3]
```diff
@@ -418,49 +420,49 @@ def _prepare(self):
...
         global_attrs = ['global_axis', 'global_tilts', 'all_bends']
         attrnames = list(self.attr_shapes.keys()) + global_attrs
         # summarise
-        self.summary = []
+        self.results.summary = []
         for i in range(len(self.atomgroups)):
             stats = {}
             for name in attrnames:
-                attr = getattr(self, name)
+                attr = getattr(self.results, name)
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/helix_analysis.py:453]: fyi
```suggestion
                attr = self.results[name]
```

File: package/MDAnalysis/analysis/helix_analysis.py [hunk 4]
```diff
@@ -418,49 +420,49 @@ def _prepare(self):
...
-            self.summary.append(stats)
+            self.results.summary.append(stats)
 
         # flatten?
         if len(self.atomgroups) == 1 and self._flatten:
             for name in attrnames + ['summary']:
-                attr = getattr(self, name)
-                setattr(self, name, attr[0])
+                attr = getattr(self.results, name)
+                setattr(self.results, name, attr[0])
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/helix_analysis.py:465]:
```suggestion
                self.results[name] = attr[0]
```

File: package/MDAnalysis/analysis/helix_analysis.py [hunk 5]
```diff
@@ -418,49 +420,49 @@ def _prepare(self):
...
                                'abs_dev': dev.mean(axis=0)}
-            self.summary.append(stats)
+            self.results.summary.append(stats)
 
         # flatten?
         if len(self.atomgroups) == 1 and self._flatten:
             for name in attrnames + ['summary']:
-                attr = getattr(self, name)
-                setattr(self, name, attr[0])
+                attr = getattr(self.results, name)
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/helix_analysis.py:464]:
```suggestion
                attr = self.results[name]
```

File: testsuite/MDAnalysisTests/analysis/test_helix_analysis.py [hunk 6]
```diff
@@ -368,11 +368,11 @@ def test_regression_values(self):
 
         for key, value in HELANAL_SINGLE_DATA.items():
             if 'summary' in key:
-                data = getattr(ha, key.split()[0])
+                data = getattr(ha.results, key.split()[0])
```
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_helix_analysis.py:371]:
```suggestion
                data = ha.results[key.split()[0]]
```

File: testsuite/MDAnalysisTests/analysis/test_helix_analysis.py [hunk 7]
```diff
@@ -368,11 +368,11 @@ def test_regression_values(self):
...
                 calculated = [data.mean(), data.std(ddof=1),
                               np.fabs(data-data.mean()).mean()]
             else:
-                calculated = getattr(ha, key)[0]
+                calculated = getattr(ha.results, key)[0]
```
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_helix_analysis.py:375]:
```suggestion
                calculated = ha.results[key][0]
```

**Review by orbeckst — APPROVED** [review]
lgtm, just suggestions how to make use of the fact that Results is a dict and get rid of ugly {get,set}attr.

**Comment by orbeckst** [comment]
Something is not right in how the results are stored now. I am not sure what, I just looked at the failures and they are typically an extra (or missing) level of `[ .. ]` in the results. Maybe I made a mistake in my quick attr <-> dict translation.

File: package/MDAnalysis/analysis/helix_analysis.py [hunk 8]
```diff
@@ -461,8 +461,8 @@ def _conclude(self):
         # flatten?
         if len(self.atomgroups) == 1 and self._flatten:
             for name in attrnames + ['summary']:
-                attr = getattr(self.results, name)
-                setattr(self.results, name, attr[0])
+                attr = self.results[name]
+                self.results[name] = attr[0]
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/helix_analysis.py:465]: Is this really
```python
self.results[name] = self.results[name][0]
```
?

Comment (lilyminium) [reply package/MDAnalysis/analysis/helix_analysis.py:465]: It is. I think this is an issue with the Results class. The below fails in Results.

```python
def __setitem__(self, key, item):
        self._validate_key(key)
        super().__setitem__(key, item)
        assert self[key] is getattr(self, key)
```

Comment (lilyminium) [reply package/MDAnalysis/analysis/helix_analysis.py:465]: See [this falsifying test](https://github.com/MDAnalysis/mdanalysis/pull/3281/commits/cb2c34edfebbc079ac530515474fe51dfd3fa44e) for an example.

**Comment by orbeckst** [comment]
I assume you need PR #3281 merged before restarting this one?

**Review by orbeckst — CHANGES_REQUESTED** [review]
I'll block the PR until the tests pass (which requires solving #3282 )

**Pull request merged**